### PR TITLE
Feature "Cancel booking"

### PR DESCRIPTION
**Pull request by sorokinvj**
User can press “cancel booking” on the bookings (pending, confirmed) only up until midnight the day before the first day of the booking.

- added Cancel booking button on BookingListPreview card
- increased margins between cards, so they would look like cards indeed
- added Spinners and state management to all async requests
- added Calculator icon from heroicons by creating it in `components/icons` by copying the svg code (let me know if you want to use npm package instead)

### CODE REVIEW

File: components/BookingListPreview.js [hunk 1]
```diff
@@ -12,23 +11,49 @@ const ListingListPreview = ({ booking }) => {
...
+      </p>
+      {booking.get('description') && (
+        <p>
+          {booking.get('description').slice(0, 120)}
+          {booking.get('description').length > 120 && '...'}
+        </p>
+      )}
+      <div className="my-2">
+        <button className="btn" onClick={cancelBooking}>
+          Cancel booking
```
Comment (closerearth) [inline components/BookingListPreview.js:52]: This should be wrapped in localization function (see above).

**Comment by closerearth** [comment]
Check out the user journey described in https://www.notion.so/traditionaldreamfactory/Booking-Flow-61fc80bae9df48eeb0bb8f3383a93154

Cancel button should open a new page (/booking/:id/cancel).

I just added the endpoints you'll need, might need some tweaking.
The cancel page should display the amount to refund based on a dynamic cancel policy which you can fetch using `/bookings/cancelation-policy`
Upon confirming that the user wishes to cancel the booking:
1. delete the Proof of Presence (coordinate with @juliencharrel to learn how to interact with blockchain)
2. call `/bookings/:id/cancel` to mark the booking as cancelled (the backend should process Stripe refund - for now it just emails admin)



Maybe a separate MR, but we should update the page to be "My bookings" - loading

**Comment by closerearth** [comment]
(@sorokinvj - and btw make sure to check the result from the Vercel bot, right now your branch isn't working)

**Comment by sorokinvj** [comment]
Thanks @closerearth, a couple questions about the api: 
1. Does it support auth? Do I need to get a token first? 

`/bookings/:id/cancel`:
1. What method shall I use? POST? Do I need to send a body or just call it? Will it have just a 2xx response?

`/bookings/cancelation-policy`:
1. GET?
2. Do i need to send booking id (or user id?) somehow?

**Comment by sorokinvj** [comment]
@closerearth what is the best flow to check for an icon needed? how do I parse `/fonts/icons/oasa-icons` to look for the `calculator icon` for the Refund Total title, `!?` icon for the cancel page title, and a `big cross` for the Booking Cancelled title?

**Comment by sorokinvj** [comment]
@closerearth here is the response from `/bookings/cancelation-policy`:
```
{
  default: 1
  lastday: 0
  lastmonth: 0.5
  lastweek: 0
}
```
what is the formula for calculating the refund total from these values?

**Comment by sorokinvj** [comment]
I think with every async request we would want to show a spinner somewhere, could be in place of a text of a button, or somewhere else. 
Also we need a place for an error on the Cancel page

**Comment by closerearth** [comment]
- You should have state management anytime you do api calls yes.

- the cancelation policy returns %.

- **icons** - those are not used (oasa-icons). We started moving to using https://heroicons.com/ - could you please create a new Icon component that just has a library of heroicons, and can be used as `<Icon type="calculator" />` or similar? (You can simply "copy jsx" from that site and add the icons we need in there). 
We were using FaIcons before but heroicons is simpler and more customizable.

- /bookings/:id/cancel is a POST request with empty body

**Comment by sorokinvj** [comment]
> * You should have state management anytime you do api calls yes.

Then we need a spinner in designs, or a generic approach for all uncertain cases, like: 
a) show spinner instead of a button that triggered async action
OR
b) show spinner in the center of the page covering the content with 0.5 opacity white overflow 


> * the cancelation policy returns %.

@closerearth Could you please provide an explicit formula based on the response above to calculate refund total? That would be much easier than me asking hundreds questions, what `lastday`, `default`, etc. means


> * **icons** - those are not used (oasa-icons). We started moving to using https://heroicons.com/ - could you please create a new Icon component that just has a library of heroicons, and can be used as `<Icon type="calculator" />` or similar? (You can simply "copy jsx" from that site and add the icons we need in there).

ok

**Comment by juliencharrel** [comment]
> * added Cancel booking button on BookingListPreview card
> * increased margins between cards, so they would look like cards indeed
> 
> What is needed now:
> 
> * to get the payment_intent or charge id, that was created with Stripe payment, so we could make a refund whenever user decided to cancel the booking
> * probably we need some confirmation after Cancel is clicked. Do we have a design for that already?
> * how do we want to change the booking record in DB? set the field "canceled", delete the record altogether? do smth else?

Why do you needed the payment intent in the front? Refund should happen from the backend

**Comment by juliencharrel** [comment]
> > * You should have state management anytime you do api calls yes.
> 
> Then we need a spinner in designs, or a generic approach for all uncertain cases, like: a) show spinner instead of a button that triggered async action OR b) show spinner in the center of the page covering the content with 0.5 opacity white overflow
> 

There's a spinner already in components (Spinner) and it's already used for blockchain calls, both out of flow (top right of the page), when you send coins from your wallet (blockchainwallet.js) OR at the last step of the booking process, it's embedded in the button and loads when user waits (checkout.js)

File: utils/helpers.js [hunk 2]
```diff
@@ -181,3 +180,17 @@ export const getSample = (field) => {
     throw new Error(`Invalid model type:${field.type}`);
   }
 }
+
+export const calculateRefundTotal = (value, policy) => {
+  // if (!policy) {
+  //   return 0;
+  // }
+  // const { refundableUntil, refundablePercent } = policy;
```
Comment (sorokinvj) [inline utils/helpers.js]: @closerearth this code snippet was suggested by Github Co-Pilot, I am leaving it here just as a proposal what `/cancellation-policy` response might include and how the refund calculation might be implemented. 

I cannot understand how to use the current response, please provide some explanation of what to do with `{default: 1,lastday: 0, lastmonth: 0.5, lastweek: 0}`.

I could also base the calculation purely on the description from the user journey, but I believe the endpoint has been created for a reason, so please advise on the best approach. 

I also think we'd better have a unit test for `calculateRefundTotal` because it is a crucial piece of the software, and it is better to test edge cases before they would happen in production. So if you are ok with that, after figuring out the implementation of the function, let's install [react testing library](https://testing-library.com/docs/react-testing-library/intro/), use some mocks and make a unit test

Comment (closerearth) [reply utils/helpers.js]: IF (daysleftbefore booking.start > 30) {
   `refund ${policy.default * 100}%`
}
ELSE IF (daysleftbefore booking.start > 7) {
   `refund ${policy.lastmonth * 100}%`
}
ELSE IF (daysleftbefore booking.start > 1) {
   `refund ${policy.lastweek * 100}%`
}
ELSE IF (daysleftbefore booking.start > 0) {
   `refund ${policy.lastday * 100}%`
}
ELSE    {
    'This booking is in the past'. (error state - cannot be cancelled)
}

File: utils/helpers.js [hunk 3]
```diff
@@ -181,3 +181,34 @@ export const getSample = (field) => {
     throw new Error(`Invalid model type:${field.type}`);
   }
 }
+
+export const calculateRefundTotal = (booking, policy) => {
+  const { price: { val: initialValue }, start, volunteer } = booking
+  if (volunteer) return 0
```
Comment (sorokinvj) [inline utils/helpers.js:204]: @closerearth here is the change I was mentioning in the chat, if a guest booked his stay as a volunteer we just return 0, let me know if that will work

Comment (closerearth) [reply utils/helpers.js:204]: That's fine for now yes

File: components/BookingListPreview.js [hunk 4]
```diff
@@ -1,34 +1,57 @@
...
 
   return (
-    <div className="booking-list-preview card">
-      <p>{ __('bookings_status') } <b>{booking.get('status')}</b></p>
-      <p>{ __('bookings_checkin') } <b>{start.format('LLL')}</b></p>
-      <p>{ __('bookings_checkout') } <b>{end.format('LLL')}</b></p>
-      <p>{ __('bookings_total') }
-        <b className={ booking.get('volunteer') ? 'line-through': '' }>
-          {' '}{priceFormat(booking.get('price'))}
+    <div className="booking-list-preview card mb-4 last:mb-0 md:w-fit">
```
Comment (closerearth) [inline components/BookingListPreview.js]: Let's not use w-fit - we want rather the size to be defined by the container.

Comment (sorokinvj) [reply components/BookingListPreview.js]: ok, lets reset it for now and in the next pr - My bookings I will fix it with the card itself @closerearth

File: utils/helpers.js [hunk 5]
```diff
@@ -196,4 +197,35 @@ export const getSample = (field) => {
...
+    return initialValue * lastday
+  }
+  return 0
+}
+
+export const getIsBookingCancellable = (bookingStart, bookingStatus) => {
+  const now = dayjs()
+  const start = dayjs(bookingStart)
+  const isBookingNotStarted = now.isBefore(start)
+  const allowedBookingStatuses = ['confirmed', 'open']
```
Comment (closerearth) [inline utils/helpers.js:229]: "pending"

File: pages/events/[slug]/edit.js [hunk 6]
```diff
@@ -41,9 +41,9 @@ const EditEvent = ({ event }) => {
           </Link>
           {__('events_slug_edit_link')} <i>{event.name}</i>
         </h1>
-        {!config.STRIPE_PUB_KEY && (
-          <div className="p2 italic">{__('events_no_stripe_integration')}</div>
-        )}
+        { !config.STRIPE_PUB_KEY &&
+          <div className="p2 italic">This platform doesn&apos;t have a global Stripe integration. If you wish to set paid tickets for your event you must define a custom Stripe integration in the tickets tab below.</div>
```
Comment (closerearth) [inline pages/events/[slug]/edit.js]: What's happening here...

Comment (sorokinvj) [reply pages/events/[slug]/edit.js]: ops, I will reset the file

File: utils/helpers.js (hunk 5 above)
Comment (sorokinvj) [reply utils/helpers.js:229]: so thats the thing - currently it is "open", not "pending"

Screenshot 2022-11-16 at 13 56 27

File: pages/events/[slug]/edit.js (hunk 6 above)
Comment (sorokinvj) [reply pages/events/[slug]/edit.js]: restored

**Comment by closerearth** [comment]
The linter must have run, you branch has 22k lines changes

**Comment by closerearth** [comment]
Although when I merge locally seems reasonable the changes.

**Review by closerearth — APPROVED** [review]
